### PR TITLE
add option to skip preprocess

### PR DESCRIPTION
**Pull request by hanw**
Add an option to skip preprocessing in p4c, to give user the flexibility to separate preprocessing from compilation. Driver may choose to call cpp directly without going through p4c.

### CODE REVIEW

File: frontends/common/parseInput.cpp [hunk 1]
```diff
@@ -22,10 +22,14 @@ limitations under the License.
...
-    FILE* in = options.preprocess();
-    if (::errorCount() > 0 || in == nullptr)
-        return nullptr;
-
+    FILE* in = nullptr;
+    if (options.doNotPreprocess) {
+        in = fopen(options.file, "r");
+    } else {
+        options.preprocess();
+        if (::errorCount() > 0 || in == nullptr)
```
Comment (mbudiu-vmw) [inline frontends/common/parseInput.cpp]: Shouldn't there be a null check after fopen too?

Comment (hanw) [inline frontends/common/parseInput.cpp]: Sorry, fixed now.

**Comment by mbudiu-vmw** [comment]
gcc does not seem to have a way to skip preprocessing by using command-line options.
However, the file suffix is used for this purpose.
What is the problem if the preprocessor is run on a file which has already been preprocessed?

**Comment by antoninbas** [comment]
@mbudiu-vmw it's a feature request by me, I want to be able to run p4c on a machine where there is no C compiler installed.

**Comment by mbudiu-vmw** [comment]
That makes sense, it has always bothered me that we have a dependence on cpp.
It would have been nice to use a "standard" flag for this purpose, but there is no such flag...
We could also use the .i convention for the file names, or maybe p4i.

File: frontends/common/options.cpp [hunk 2]
```diff
@@ -54,6 +54,9 @@ CompilerOptions::CompilerOptions() : Util::Options(defaultMessage) {
     registerOption("-E", nullptr,
                    [this](const char*) { doNotCompile = true; return true; },
                    "Preprocess only, do not compile (prints program on stdout)");
+    registerOption("-e", nullptr,
+                   [this](const char*) { doNotPreprocess = true; return true; },
```
Comment (ChrisDodd) [inline frontends/common/options.cpp]: might be better to use a long option like `--nocpp`?

**Pull request merged**